### ISSUE TITLE
Reload doesn't work in debug mode when code contains the `threading.Timer()` function in it. 

### ISSUE THREAD

**Issue by ahadrt** [Kludex/uvicorn#932]
## The problem
My code was able to debug and was reloading when code was changed. After adding code which contained the `threading.Timer()` function stops after code change in every file in project. It frustrated  me when I had to start server again at every file change. It seems to be bug and if any one have a solution or experience, I will happy if helps me.

after file change it shows this result.
```
INFO:     Application shutdown complete.
INFO:     Finished server process [15791]
```

#### The code sample which causes this behavior is just like this.
```python

    def remove_expired_sessions(self):
            threading.Timer(1200, self.remove_expired_sessions).start()
```

I found a way in order bypass this problem by adding a environment variable in `os.environment` and check it in the code that causes this issue. However it prevent to debug the particular part of code which contains `Timer`.

#### My temporary solution is like codes below.

In server file:

```python

 if mode == "development":
        import os
        os.environ['mode'] = 'development'
        uvicorn.run("server:app", host=host, port=22659, debug=True, reload=True, headers=[("some", "other")])


```

in module:
```python
    def remove_expired_sessions(self):
        if os.environ.get('mode') != 'development':
            threading.Timer(1200, self.remove_expired_sessions).start()

```

### Test platform and environment:
Python version: 3.8.5
vicorn version: 0.11.8
App: FastApi vesrion 0.61.1
OS: MacOs BigSur 11.1
Hardware: mac book pro 2018 core i9

**Comment by ahadrt** [comment]
In addition this issue does not happen in windows platform!

**Comment by euri10** [comment]
couple of things @ahadrt : this is an old uvicorn version, do you have same issue with latest one ? Also when you press the Issue button, the bug report section asks you to to fill couple things to help us help you, in the current report there's way too much to guess to be able to reproduce.

**Comment by ahadrt** [comment]
Dear @euri10: 
I upgraded to latest version of uvicorn (0.13.3) and fastapi to (0.63.0) as well, however it resulted the same. It seems to be a clear problem, despite this, if any additional information is needed I will provide it.

**Comment by max-l** [comment]
I can confirm I'm having the problem with version 0.13.4

**Comment by euri10** [comment]
Unless you or OP posts a minimal reproducible example there's little chance to make progress.

**Comment by euri10** [comment]
closing for now, still lacks info to reproduce, feel free to reopen with it

**Comment by kfdl** [comment]
Hello, 

I have the same issue on my side, 
Do you have any update, that will be very appreciate

thank :)

**Comment by ahadrt** [comment]
@kfdl Hi 
I realized that problem was the workers which I'd set to uvicorn. I tried to make a  in memory singleton session manager to hold some data when a user connects to server. however I it failed not because of workers and separation between processes. In fact singleton pattern cant be used in multiprocess pattern because, processes are not able to access to each-other's memory, then every singleton object will be instantiate separately in each process. I solved my problem by changing memory based session manager to database based session. In this case all processes call the same session data from single database.
Long story short, my expectation was wrong.

**Comment by FarhadG** [comment]
@ahadrt I believe I'm having a very similar issue as you. I have an internal memory store with a background process to clear items that are expired (based on a TTL). During a reload, my service is hanging which requires for me to restart the entire application. Can you please provide more detail as to how you solved the issue?  I'd like to get a better understanding as it seems you had a similar implementation.
